### ISSUE TITLE
List_nonempty module

### ISSUE THREAD

**Issue by rixed** [ocaml-batteries-team/batteries-included#1057]
from https://discuss.ocaml.org/t/share-your-crazy-ocaml-code-snippet/8116/38 :

```
Sometimes, I think a feature-full List_nonempty module would be cool in batteries, or as a standalone library.
It should support most BatList operations (and only have tail rec. code).
```

I too think it would be useful.

**Comment by UnixJunkie** [comment]
Apparently there is one in Core_kernel: https://github.com/janestreet/core_kernel/tree/master/nonempty_list
We could import part of it and use the interface we want (though this implementation file would be using the MIT license then).

**Comment by UnixJunkie** [comment]
One way to start resolving this issue would be to agree on the interface, which I believe should be a subset of batList.
Once the interface is agreed upon, we could implement all the functions.

**Comment by rixed** [comment]
Indeed, I see no reason why we should not aim for a subset of BatList as far as API goes.
What seams less obvious to me is what should be the actual encoding. Should a non empty list be an item + a normal list (as in Core's implementation it seems), which has the advantage of offering a free conversion from/to lists ; or should a list be either the last item or an item + a non empty list, which seems marginally cleaner and makes the definition "safer" (it can then be public and allow users to pattern match directly).

**Comment by UnixJunkie** [comment]
If the data structure definition is like Core's one, the advantage I see is
that the implementation
for many functions is pretty much just calling the corresponding function
from BatList.
And since BatList has a lot of functions, that would be nice to
simplify/speedup the implementation.

**Comment by rixed** [comment]
This makes a lot of sense.

**Comment by UnixJunkie** [comment]
I have started some work here: https://github.com/UnixJunkie/batteries-included/tree/non_empty_list
The idea is to uncomment little by little the bottom of the file, implementing things along the way;
keeping BatList's unit tests as much as possible
